pkg/extensions/tbtc/recovery/storage.go

```diff
@@ -0,0 +1,135 @@
+package recovery
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/keep-network/keep-common/pkg/encryption"
+)
+
+const (
+	fileName = "derivation_indexes"
+)
+
+type DerivationIndexStorage struct {
+	path string
+	box  encryption.Box
+}
+
+func NewDerivationIndexStorage(path string, password string) *DerivationIndexStorage {
+	passwordBytes := []byte(password)
+
+	return &DerivationIndexStorage{
+		path: path,
+		box:  encryption.NewBox(sha256.Sum256(passwordBytes)),
+	}
+}
+
+func (dis DerivationIndexStorage) getStoragePath() string {
+	return fmt.Sprintf("%s/%s", dis.path, fileName)
+}
+
+func (dis DerivationIndexStorage) Save(extendedPublicKey string, index int) error {
+	dirPath := dis.getStoragePath()
+
+	addressIndexes, err := dis.readMap()
+	if err != nil {
+		return err
+	}
+
+	addressIndexes[extendedPublicKey] = index
+	json, err := json.Marshal(addressIndexes)
+	if err != nil {
+		return err
+	}
+	encrypted, err := dis.box.Encrypt(json)
+	if err != nil {
+		return nil
+	}
+
+	return write(dirPath, encrypted)
+}
+
+func (dis DerivationIndexStorage) Read(extendedPublicKey string) (int, error) {
+	addressIndexes, err := dis.readMap()
+	if err != nil {
+		return 0, err
+	}
+	index, ok := addressIndexes[extendedPublicKey]
+	if ok {
+		return index, nil
+	}
+	return 0, nil
+}
+
+func (dis DerivationIndexStorage) readMap() (map[string]int, error) {
```

Comment:
I like it! Do you want the files containing the btc addresses to be encrypted?

Reply:
Actually, I've just checked the RFC and it says that:

> To avoid strictly requiring a connection to a Bitcoin node, this RFC proposes tracking the latest address generated, and allowing an additional config key, TargetDerivationStartIndex, that sets a starting point for such derivations. The client should write the updated index to the current/ directory that tracks active key shares, to an encrypted file named current-btc-refund-derivation-index. The operator should then seek to obtain from the staker an extended public key that is only used for this purpose. In cases where in-memory/on-disk derivation index and config file index both exist, the highest index should be used.

I see two problems with it though. `ReadAll` in the storage will try to load `current-btc-refund-derivation-index` file as a threshold signer and will explode. The second problem is what happens if the operator changes their BTC public key in the config? Do we use the existing derivation index or do we start from 0? I'd say starting from 0 makes more sense.

If we ignore the starting index issue, we could update the existing `persistentStorage` to do not read anything that [is not a valid address](https://github.com/keep-network/keep-ecdsa/blob/693068b463ca9268fda8732dab73bb750ff60dd0/pkg/registry/storage.go#L120-L126) and add a separate function next to `readRaw` that would read specific file bytes without trying to deserialize it to `tss.ThresholdSigner`. We'd also need `saveRaw`. 

I think the RFC aimed at least-effort implementation and honestly, I think that what you are doing here is even less effort than trying to adapt `persistentStorage` the way we want it to work.

Reply:
Good question on encryption. I... don't think it needs to be encrypted. The same information can be evaluated just by looking at the config file 🤔 I think the encryption piece comes from the fact we aimed at least-effort implementation. Did I miss anything?

Reply:
That's what I was wondering - if the extended public key is in plain text in the config, and the way the derivation path stuff works is open source, I don't see us *currently* protecting folks from figuring out who owns which btc addresses if someone manages to get filesystem access to the box